Samyuktha Chaparla

In [233]:
# Generic inputs for most ML tasks
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
# This is new
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.ensemble import RandomForestRegressor

pd.options.display.float_format = '{:,.2f}'.format

# setup interactive notebook mode
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

from IPython.display import display, HTML

In [234]:
df= pd.read_csv('invistico_Airline.csv')
df.head()

satisfaction  Gender   Customer Type  Age   Type of Travel     Class  \
0    satisfied  Female  Loyal Customer   65  Personal Travel       Eco   
1    satisfied    Male  Loyal Customer   47  Personal Travel  Business   
2    satisfied  Female  Loyal Customer   15  Personal Travel       Eco   
3    satisfied  Female  Loyal Customer   60  Personal Travel       Eco   
4    satisfied  Female  Loyal Customer   70  Personal Travel       Eco   

   Flight Distance  Seat comfort  Departure/Arrival time convenient  \
0              265             0                                  0   
1             2464             0                                  0   
2             2138             0                                  0   
3              623             0                                  0   
4              354             0                                  0   

   Food and drink  ...  Online support  Ease of Online booking  \
0               0  ...               2                       3   
1               0  ...               2                       3   
2               0  ...               2                       2   
3               0  ...               3                       1   
4               0  ...               4                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 3                 0                 3                5   
1                 4                 4                 4                2   
2                 3                 3                 4                4   
3                 1                 0                 1                4   
4                 2                 0                 2                4   

   Cleanliness  Online boarding  Departure Delay in Minutes  \
0            3                2                           0   
1            3                2                         310   
2            4                2                           0   
3            1                3                           0   
4            2                5                           0   

   Arrival Delay in Minutes  
0                      0.00  
1                    305.00  
2                      0.00  
3                      0.00  
4                      0.00  

[5 rows x 23 columns]

In [235]:
df.drop(columns = 'satisfaction', inplace = True)
subset_data = df.copy()
subset_data.head()

Gender   Customer Type  Age   Type of Travel     Class  Flight Distance  \
0  Female  Loyal Customer   65  Personal Travel       Eco              265   
1    Male  Loyal Customer   47  Personal Travel  Business             2464   
2  Female  Loyal Customer   15  Personal Travel       Eco             2138   
3  Female  Loyal Customer   60  Personal Travel       Eco              623   
4  Female  Loyal Customer   70  Personal Travel       Eco              354   

   Seat comfort  Departure/Arrival time convenient  Food and drink  \
0             0                                  0               0   
1             0                                  0               0   
2             0                                  0               0   
3             0                                  0               0   
4             0                                  0               0   

   Gate location  ...  Online support  Ease of Online booking  \
0              2  ...               2                       3   
1              3  ...               2                       3   
2              3  ...               2                       2   
3              3  ...               3                       1   
4              3  ...               4                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 3                 0                 3                5   
1                 4                 4                 4                2   
2                 3                 3                 4                4   
3                 1                 0                 1                4   
4                 2                 0                 2                4   

   Cleanliness  Online boarding  Departure Delay in Minutes  \
0            3                2                           0   
1            3                2                         310   
2            4                2                           0   
3            1                3                           0   
4            2                5                           0   

   Arrival Delay in Minutes  
0                      0.00  
1                    305.00  
2                      0.00  
3                      0.00  
4                      0.00  

[5 rows x 22 columns]

Dropped the 'satisfaction' column

In [236]:
subset_data.isna().sum()

Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Seat comfort                           0
Departure/Arrival time convenient      0
Food and drink                         0
Gate location                          0
Inflight wifi service                  0
Inflight entertainment                 0
Online support                         0
Ease of Online booking                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Cleanliness                            0
Online boarding                        0
Departure Delay in Minutes             0
Arrival Delay in Minutes             393
dtype: int64

Checking for Null values and dropping them if there are any present. We ccan notice that Arrival Delay in Minutes has 393 Null values. So, in the next step we are dropping those.

In [237]:
subset_data.dropna(inplace=True)
subset_data.head()
len(subset_data)

Gender   Customer Type  Age   Type of Travel     Class  Flight Distance  \
0  Female  Loyal Customer   65  Personal Travel       Eco              265   
1    Male  Loyal Customer   47  Personal Travel  Business             2464   
2  Female  Loyal Customer   15  Personal Travel       Eco             2138   
3  Female  Loyal Customer   60  Personal Travel       Eco              623   
4  Female  Loyal Customer   70  Personal Travel       Eco              354   

   Seat comfort  Departure/Arrival time convenient  Food and drink  \
0             0                                  0               0   
1             0                                  0               0   
2             0                                  0               0   
3             0                                  0               0   
4             0                                  0               0   

   Gate location  ...  Online support  Ease of Online booking  \
0              2  ...               2                       3   
1              3  ...               2                       3   
2              3  ...               2                       2   
3              3  ...               3                       1   
4              3  ...               4                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 3                 0                 3                5   
1                 4                 4                 4                2   
2                 3                 3                 4                4   
3                 1                 0                 1                4   
4                 2                 0                 2                4   

   Cleanliness  Online boarding  Departure Delay in Minutes  \
0            3                2                           0   
1            3                2                         310   
2            4                2                           0   
3            1                3                           0   
4            2                5                           0   

   Arrival Delay in Minutes  
0                      0.00  
1                    305.00  
2                      0.00  
3                      0.00  
4                      0.00  

[5 rows x 22 columns]

129487

The total number of rows of data available after dropping Null values are 129487

In [238]:
subset_data.isna().sum()

Gender                               0
Customer Type                        0
Age                                  0
Type of Travel                       0
Class                                0
Flight Distance                      0
Seat comfort                         0
Departure/Arrival time convenient    0
Food and drink                       0
Gate location                        0
Inflight wifi service                0
Inflight entertainment               0
Online support                       0
Ease of Online booking               0
On-board service                     0
Leg room service                     0
Baggage handling                     0
Checkin service                      0
Cleanliness                          0
Online boarding                      0
Departure Delay in Minutes           0
Arrival Delay in Minutes             0
dtype: int64

In [239]:
print(set(subset_data['Gender']))
print(set(subset_data['Type of Travel']))
print(set(subset_data['Customer Type']))

{'Male', 'Female'}
{'Personal Travel', 'Business travel'}
{'Loyal Customer', 'disloyal Customer'}


Applying OneHotEncoder on the catogorical independent variables i.e., Gender, Customer Type, Type of Travel.

In [240]:
from sklearn.preprocessing import OneHotEncoder

def get_ohe(df, col):
    ohe = OneHotEncoder(drop='first', handle_unknown='error', sparse_output=False, dtype='int')
    ohe.fit(df[[col]])
    temp_df = pd.DataFrame(data=ohe.transform(df[[col]]), columns=ohe.get_feature_names_out())
    df.drop(columns=[col], axis=1, inplace=True)
    df = pd.concat([df.reset_index(drop=True), temp_df], axis=1)
    return df

In [241]:
subset_data = get_ohe(subset_data, 'Gender')
subset_data = get_ohe(subset_data, 'Customer Type')
subset_data = get_ohe(subset_data, 'Type of Travel')
subset_data.head()

Age     Class  Flight Distance  Seat comfort  \
0   65       Eco              265             0   
1   47  Business             2464             0   
2   15       Eco             2138             0   
3   60       Eco              623             0   
4   70       Eco              354             0   

   Departure/Arrival time convenient  Food and drink  Gate location  \
0                                  0               0              2   
1                                  0               0              3   
2                                  0               0              3   
3                                  0               0              3   
4                                  0               0              3   

   Inflight wifi service  Inflight entertainment  Online support  ...  \
0                      2                       4               2  ...   
1                      0                       2               2  ...   
2                      2                       0               2  ...   
3                      3                       4               3  ...   
4                      4                       3               4  ...   

   Leg room service  Baggage handling  Checkin service  Cleanliness  \
0                 0                 3                5            3   
1                 4                 4                2            3   
2                 3                 4                4            4   
3                 0                 1                4            1   
4                 0                 2                4            2   

   Online boarding  Departure Delay in Minutes  Arrival Delay in Minutes  \
0                2                           0                      0.00   
1                2                         310                    305.00   
2                2                           0                      0.00   
3                3                           0                      0.00   
4                5                           0                      0.00   

   Gender_Male  Customer Type_disloyal Customer  \
0            0                                0   
1            1                                0   
2            0                                0   
3            0                                0   
4            0                                0   

   Type of Travel_Personal Travel  
0                               1  
1                               1  
2                               1  
3                               1  
4                               1  

[5 rows x 22 columns]

In [242]:
print(f"Number of Independent variables: {subset_data.shape[1] - 1}")

Number of Independent variables: 21


we have 21 independent variables in the dataset after one-hot encoding.

Splitting the data into test and train by stratify 'Class' and with random_state as 50.  

In [243]:
X_train, X_test, y_train, y_test = train_test_split(subset_data.drop(columns = ['Class']), subset_data['Class'], test_size=0.20, stratify=subset_data['Class'],random_state=50)

X_train
X_test
y_train
y_test
first_train_index = X_train.index[0]
print(f"The start index for train dataset is {first_train_index}")
first_test_index = X_test.index[0]
print(f"The start index for test dataset is {first_test_index}")

Age  Flight Distance  Seat comfort  Departure/Arrival time convenient  \
100897   59             2559             1                                  1   
76456    37              520             2                                  3   
44676    26             1487             2                                  0   
37815    50              550             4                                  5   
122902   39             1980             5                                  1   
...     ...              ...           ...                                ...   
106779   46             2035             3                                  3   
67611    54              269             1                                  3   
119841   58              101             4                                  4   
119381   55              309             1                                  1   
111902   30             5585             3                                  3   

        Food and drink  Gate location  Inflight wifi service  \
100897               1              1                      5   
76456                3              3                      5   
44676                2              3                      3   
37815                4              4                      1   
122902               5              5                      5   
...                ...            ...                    ...   
106779               3              3                      3   
67611                4              3                      2   
119841               4              4                      5   
119381               1              1                      2   
111902               3              3                      4   

        Inflight entertainment  Online support  Ease of Online booking  ...  \
100897                       4               5                       4  ...   
76456                        4               4                       2  ...   
44676                        2               3                       3  ...   
37815                        5               5                       5  ...   
122902                       5               5                       5  ...   
...                        ...             ...                     ...  ...   
106779                       4               5                       4  ...   
67611                        3               4                       1  ...   
119841                       3               4                       5  ...   
119381                       4               4                       5  ...   
111902                       5               4                       4  ...   

        Leg room service  Baggage handling  Checkin service  Cleanliness  \
100897                 4                 4                3            4   
76456                  2                 2                1            2   
44676                  2                 4                4            5   
37815                  4                 5                3            5   
122902                 3                 5                5            1   
...                  ...               ...              ...          ...   
106779                 4                 4                5            4   
67611                  1                 1                3            1   
119841                 5                 5                5            5   
119381                 5                 5                3            5   
111902                 4                 4                3            4   

        Online boarding  Departure Delay in Minutes  Arrival Delay in Minutes  \
100897                4                           0                      0.00   
76456                 3                           0                      0.00   
44676                 3                           0                      0.00   
37815                 1                           0                      0.00   

Age  Flight Distance  Seat comfort  Departure/Arrival time convenient  \
42248    32             1727             1                                  1   
102458   57             1889             5                                  5   
236      58              622             0                                  3   
74922    41             2506             2                                  2   
14632    40             3669             4                                  3   
...     ...              ...           ...                                ...   
102243   64              256             4                                  5   
58476    26             2130             4                                  4   
44093    26             2154             1                                  4   
71815    72             1111             2                                  1   
92998    40             1606             4                                  4   

        Food and drink  Gate location  Inflight wifi service  \
42248                1              3                      5   
102458               5              5                      3   
236                  0              4                      3   
74922                2              2                      4   
14632                3              3                      4   
...                ...            ...                    ...   
102243               5              5                      4   
58476                4              1                      2   
44093                1              4                      5   
71815                1              1                      5   
92998                4              4                      4   

        Inflight entertainment  Online support  Ease of Online booking  ...  \
42248                        1               5                       5  ...   
102458                       5               4                       4  ...   
236                          1               4                       1  ...   
74922                        4               4                       2  ...   
14632                        4               4                       4  ...   
...                        ...             ...                     ...  ...   
102243                       3               4                       4  ...   
58476                        4               2                       2  ...   
44093                        1               2                       5  ...   
71815                        3               2                       2  ...   
92998                        4               4                       4  ...   

        Leg room service  Baggage handling  Checkin service  Cleanliness  \
42248                  3                 4                3            3   
102458                 4                 4                4            4   
236                    0                 1                3            1   
74922                  2                 2                4            2   
14632                  4                 4                5            1   
...                  ...               ...              ...          ...   
102243                 4                 4                2            4   
58476                  4                 5                5            4   
44093                  4                 3                3            3   
71815                  2                 2                1            2   
92998                  4                 4                3            4   

        Online boarding  Departure Delay in Minutes  Arrival Delay in Minutes  \
42248                 5                           0                      0.00   
102458                5                          75                     77.00   
236                   3                           0                      0.00   
74922                 4                           0                      0.00   

100897    Business
76456     Business
44676     Business
37815          Eco
122902         Eco
            ...   
106779    Business
67611     Business
119841    Business
119381    Business
111902    Business
Name: Class, Length: 103589, dtype: object

42248     Business
102458    Business
236            Eco
74922     Business
14632          Eco
            ...   
102243    Eco Plus
58476     Business
44093     Business
71815          Eco
92998     Business
Name: Class, Length: 25898, dtype: object

The start index for train dataset is 100897
The start index for test dataset is 42248


In [244]:
model = LogisticRegression(fit_intercept = True, solver='lbfgs', multi_class = 'ovr', penalty =None,max_iter = 100)

model.fit(X_train, y_train) 

# The following gives the mean accuracy on the given data and labels
model.score(X_train, y_train) 

# This is the coefficient Beta_1, ..., Beta_7
model.coef_

# This is the coefficient Beta_0
model.intercept_

C:\Users\samyu\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(multi_class='ovr', penalty=None)

0.633686974485708

array([[-2.76473398e-03,  1.70801741e-04, -2.34979064e-01,
        -2.58870365e-01,  4.62195733e-02, -1.33697007e-01,
        -1.43932137e-01,  3.85061801e-01,  1.53362984e-01,
         4.44993698e-02,  1.39637118e-01,  3.51886998e-02,
        -2.22905932e-02,  4.53226383e-02, -4.89657378e-02,
        -6.28008523e-02, -1.27054533e-02,  1.41416021e-02,
        -5.20349255e-02, -1.01339272e-01, -6.17207139e-01],
       [-4.75721326e-04, -2.87430382e-04,  2.21321414e-01,
         2.25941671e-01, -7.88031868e-02,  1.04988069e-01,
         1.44713896e-01, -3.98926879e-01, -1.49529288e-01,
        -3.98176741e-02, -1.23880955e-01, -3.41219515e-02,
         2.99647565e-02, -2.55685492e-02,  5.51527182e-02,
         7.46006779e-02,  1.26907359e-02, -2.64750786e-03,
         5.42124448e-02,  1.47641314e-01,  6.79485501e-01],
       [-8.90145456e-03, -2.83000658e-04,  1.86865770e-01,
         7.90360813e-02, -5.11587852e-02, -9.70446377e-02,
         9.91065427e-03, -1.83821406e-01, -1.04114307e

array([-0.11498489,  0.11069081, -0.03204597])

The score is 0.6336
The beta 0 is -0.1149
Beta 1 is 0.002764
Beta 2 is -0.0004757


In [245]:

model = LogisticRegression(fit_intercept = True, solver='lbfgs', multi_class = 'ovr', penalty = None,max_iter =1000)

model.fit(X_train, y_train) 

# The following gives the mean accuracy on the given data and labels
print(f"The model score")
model.score(X_train, y_train) 

# This is the coefficient Beta_1, ..., Beta_7
model.coef_

# This is the coefficient Beta_0
model.intercept_

C:\Users\samyu\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000, multi_class='ovr', penalty=None)

The model score


0.767504271689079

array([[-2.11164665e-03,  1.77643987e-04, -4.28167132e-01,
         9.66480973e-02,  8.42833805e-02, -6.76941540e-02,
        -4.31978318e-02,  2.97563833e-01,  1.50439526e-01,
        -1.10830309e-01,  1.89374311e-01,  3.03529972e-02,
         1.44645995e-01,  1.60699999e-01,  1.18558981e-01,
        -2.94904370e-02,  2.06653894e-03, -4.13523832e-03,
        -1.96548460e-02, -1.31343068e+00, -4.06013087e+00],
       [-1.90707451e-03, -2.01555320e-04,  2.79268375e-01,
        -5.22225055e-02, -3.22380048e-02,  3.77495093e-03,
         3.92272534e-02, -2.28045049e-01, -1.47556778e-01,
         1.32732787e-01, -1.39180783e-01, -4.19367989e-02,
        -1.03163854e-01, -1.17853450e-01, -1.02141779e-01,
        -7.81547251e-03,  5.77190595e-04,  1.08226792e-03,
        -4.02064442e-03,  1.51359116e+00,  3.16655219e+00],
       [-7.04147100e-03, -2.04574208e-04,  1.83986611e-01,
        -5.30450129e-03, -2.30567612e-02, -4.90423796e-02,
         4.70662092e-03, -1.92210398e-01, -1.07377597e

array([-1.2725709 ,  0.84855056, -0.12261946])

In [246]:
test_output = pd.DataFrame(model.predict(X_test), index = X_test.index, columns = ['pred_Class'])
test_output.head()

pred_Class
42248         Eco
102458   Business
236           Eco
74922    Business
14632         Eco

In [247]:
test_output = test_output.merge(y_test, left_index = True, right_index = True)
test_output.head()
print('Percentage of correct predictions is ')
score_rounded = model.score(X_test, y_test)
print(score_rounded)

pred_Class     Class
42248         Eco  Business
102458   Business  Business
236           Eco       Eco
74922    Business  Business
14632         Eco       Eco

Percentage of correct predictions is 
0.7686307822997915


In [248]:
if True:
    from sklearn.preprocessing import StandardScaler
    sc = StandardScaler()
    X_train = pd.DataFrame(sc.fit_transform(X_train), columns = X_train.columns, index = X_train.index)
    X_test = pd.DataFrame(sc.transform(X_test), columns = X_test.columns, index = X_test.index)
    X_train
    X_test
    y_train
    y_test

Age  Flight Distance  Seat comfort  \
100897  1.29             0.56         -1.32   
76456  -0.16            -1.42         -0.60   
44676  -0.89            -0.48         -0.60   
37815   0.70            -1.39          0.84   
122902 -0.03            -0.00          1.55   
...      ...              ...           ...   
106779  0.44             0.05          0.12   
67611   0.96            -1.67         -1.32   
119841  1.23            -1.83          0.84   
119381  1.03            -1.63         -1.32   
111902 -0.62             3.51          0.12   

        Departure/Arrival time convenient  Food and drink  Gate location  \
100897                              -1.30           -1.28          -1.52   
76456                                0.01            0.10           0.01   
44676                               -1.95           -0.59           0.01   
37815                                1.32            0.80           0.78   
122902                              -1.30            1.49           1.54   
...                                   ...             ...            ...   
106779                               0.01            0.10           0.01   
67611                                0.01            0.80           0.01   
119841                               0.66            0.80           0.78   
119381                              -1.30           -1.28          -1.52   
111902                               0.01            0.10           0.01   

        Inflight wifi service  Inflight entertainment  Online support  \
100897                   1.33                    0.46            1.13   
76456                    1.33                    0.46            0.37   
44676                   -0.19                   -1.03           -0.40   
37815                   -1.71                    1.20            1.13   
122902                   1.33                    1.20            1.13   
...                       ...                     ...             ...   
106779                  -0.19                    0.46            1.13   
67611                   -0.95                   -0.29            0.37   
119841                   1.33                   -0.29            0.37   
119381                  -0.95                    0.46            0.37   
111902                   0.57                    1.20            0.37   

        Ease of Online booking  ...  Leg room service  Baggage handling  \
100897                    0.41  ...              0.40              0.26   
76456                    -1.13  ...             -1.15             -1.47   
44676                    -0.36  ...             -1.15              0.26   
37815                     1.17  ...              0.40              1.13   
122902                    1.17  ...             -0.38              1.13   
...                        ...  ...               ...               ...   
106779                    0.41  ...              0.40              0.26   
67611                    -1.89  ...             -1.92             -2.34   
119841                    1.17  ...              1.17              1.13   
119381                    1.17  ...              1.17              1.13   
111902                    0.41  ...              0.40              0.26   

        Checkin service  Cleanliness  Online boarding  \
100897            -0.27         0.26             0.50   
76456             -1.86        -1.48            -0.27   
44676              0.52         1.12            -0.27   
37815             -0.27         1.12            -1.81   
122902             1.32        -2.35             1.27   
...                 ...          ...              ...   
106779             1.32         0.26            -0.27   
67611             -0.27        -2.35             0.50   
119841             1.32         1.12             0.50   
119381            -0.27         1.12             1.27   
111902            -0.27         0.26             0.50   

        Departure Delay in Minutes  Arrival Delay in Minutes  Gender

Age  Flight Distance  Seat comfort  \
42248  -0.49            -0.25         -1.32   
102458  1.16            -0.09          1.55   
236     1.23            -1.32         -2.04   
74922   0.10             0.51         -0.60   
14632   0.04             1.64          0.84   
...      ...              ...           ...   
102243  1.63            -1.68          0.84   
58476  -0.89             0.14          0.84   
44093  -0.89             0.17         -1.32   
71815   2.15            -0.85         -0.60   
92998   0.04            -0.37          0.84   

        Departure/Arrival time convenient  Food and drink  Gate location  \
42248                               -1.30           -1.28           0.01   
102458                               1.32            1.49           1.54   
236                                  0.01           -1.98           0.78   
74922                               -0.65           -0.59          -0.76   
14632                                0.01            0.10           0.01   
...                                   ...             ...            ...   
102243                               1.32            1.49           1.54   
58476                                0.66            0.80          -1.52   
44093                                0.66           -1.28           0.78   
71815                               -1.30           -1.28          -1.52   
92998                                0.66            0.80           0.78   

        Inflight wifi service  Inflight entertainment  Online support  \
42248                    1.33                   -1.77            1.13   
102458                  -0.19                    1.20            0.37   
236                     -0.19                   -1.77            0.37   
74922                    0.57                    0.46            0.37   
14632                    0.57                    0.46            0.37   
...                       ...                     ...             ...   
102243                   0.57                   -0.29            0.37   
58476                   -0.95                    0.46           -1.16   
44093                    1.33                   -1.77           -1.16   
71815                    1.33                   -0.29           -1.16   
92998                    0.57                    0.46            0.37   

        Ease of Online booking  ...  Leg room service  Baggage handling  \
42248                     1.17  ...             -0.38              0.26   
102458                    0.41  ...              0.40              0.26   
236                      -1.89  ...             -2.70             -2.34   
74922                    -1.13  ...             -1.15             -1.47   
14632                     0.41  ...              0.40              0.26   
...                        ...  ...               ...               ...   
102243                    0.41  ...              0.40              0.26   
58476                    -1.13  ...              0.40              1.13   
44093                     1.17  ...              0.40             -0.60   
71815                    -1.13  ...             -1.15             -1.47   
92998                     0.41  ...              0.40              0.26   

        Checkin service  Cleanliness  Online boarding  \
42248             -0.27        -0.61             1.27   
102458             0.52         0.26             1.27   
236               -0.27        -2.35            -0.27   
74922              0.52        -1.48             0.50   
14632              1.32        -2.35             0.50   
...                 ...          ...              ...   
102243            -1.06         0.26            -1.04   
58476              1.32         0.26            -1.04   
44093             -0.27        -0.61             1.27   
71815             -1.86        -1.48            -0.27   
92998             -0.27         0.26             1.27   

        Departure Delay in Minutes  Arrival Delay in Minutes  Gender

100897    Business
76456     Business
44676     Business
37815          Eco
122902         Eco
            ...   
106779    Business
67611     Business
119841    Business
119381    Business
111902    Business
Name: Class, Length: 103589, dtype: object

42248     Business
102458    Business
236            Eco
74922     Business
14632          Eco
            ...   
102243    Eco Plus
58476     Business
44093     Business
71815          Eco
92998     Business
Name: Class, Length: 25898, dtype: object

In [249]:
model = LogisticRegression(fit_intercept = True, solver='lbfgs', multi_class = 'ovr', penalty = None,max_iter =1000)

model.fit(X_train, y_train) 

# The following gives the mean accuracy on the given data and labels
print(f"The model score")
model.score(X_train, y_train) 

# This is the coefficient Beta_1, ..., Beta_7
model.coef_

# This is the coefficient Beta_0
model.intercept_

LogisticRegression(max_iter=1000, multi_class='ovr', penalty=None)

The model score


0.7722538107327999

array([[ 1.92350151e-01,  3.98843275e-01, -7.48912432e-01,
         7.38734817e-02,  2.24779006e-01,  3.63847633e-02,
         5.92900411e-02,  5.73412810e-01,  3.13650234e-01,
        -3.81258474e-01,  3.51603750e-01,  1.21784845e-01,
         2.93748295e-01,  2.64173994e-01,  2.66793956e-01,
         1.51422229e-02,  1.05983253e-01, -1.67093697e-01,
         2.44247769e-02, -5.20854523e-01, -2.01951663e+00],
       [-1.23328479e-01, -3.00680892e-01,  4.94654072e-01,
        -3.00284129e-02, -1.69629613e-01, -3.01794344e-02,
        -5.13443440e-02, -3.71721706e-01, -2.14756365e-01,
         2.84019325e-01, -2.30218397e-01, -9.80804454e-02,
        -2.05483289e-01, -1.86622955e-01, -1.98943371e-01,
        -2.87800670e-02, -6.43083629e-02,  1.16245397e-01,
        -1.69963887e-02,  6.18625698e-01,  1.50362227e+00],
       [-1.01247359e-01, -2.05972879e-01,  2.72977991e-01,
         2.16801420e-03, -5.29509488e-02, -6.50657223e-02,
         1.02114691e-03, -2.59700973e-01, -1.38714436e

array([-0.28500927, -0.3125446 , -2.73450827])

In [250]:
test_output = pd.DataFrame(model.predict(X_test), index = X_test.index, columns = ['pred_Class'])
test_output.head()

pred_Class
42248         Eco
102458   Business
236           Eco
74922    Business
14632         Eco

In [251]:
test_output = test_output.merge(y_test, left_index = True, right_index = True)
test_output.head()
print('Percentage of correct predictions is ')
score_rounded = model.score(X_test, y_test)
print(score_rounded)

pred_Class     Class
42248         Eco  Business
102458   Business  Business
236           Eco       Eco
74922    Business  Business
14632         Eco       Eco

Percentage of correct predictions is 
0.775040543671326


In [252]:
incorrect_predictions = test_output[test_output['pred_Class'] != test_output['Class']]
incorrect_counts = incorrect_predictions['Class'].value_counts()
most_wrong_class = incorrect_counts.idxmax()
fewest_wrong_class = incorrect_counts.idxmin()

print("Most wrong class:", most_wrong_class)
print("Fewest wrong class:", fewest_wrong_class)

Most wrong class: Eco
Fewest wrong class: Business


In [253]:
model = LogisticRegression(fit_intercept = True, solver='newton-cg', multi_class = 'ovr', penalty = None, max_iter = 1000)
model.fit(X_train, y_train) 

# The following gives the mean accuracy on the given data and labels
print(f"The model score")
model.score(X_train, y_train) 

# This is the coefficient Beta_1, ..., Beta_7
model.coef_

# This is the coefficient Beta_0
model.intercept_

LogisticRegression(max_iter=1000, multi_class='ovr', penalty=None,
                   solver='newton-cg')

The model score


0.7722731178020832

array([[ 1.91428572e-01,  3.98505996e-01, -7.49240103e-01,
         7.38328088e-02,  2.25381402e-01,  3.60125748e-02,
         5.92655255e-02,  5.73651664e-01,  3.13759109e-01,
        -3.81607026e-01,  3.51533692e-01,  1.21769845e-01,
         2.93965978e-01,  2.64281757e-01,  2.66967412e-01,
         1.51644631e-02,  1.06233550e-01, -1.66754309e-01,
         2.45008985e-02, -5.21369250e-01, -2.01956171e+00],
       [-1.23242029e-01, -3.00740438e-01,  4.95017351e-01,
        -2.97651329e-02, -1.70800756e-01, -2.99263401e-02,
        -5.07906506e-02, -3.71233100e-01, -2.14874627e-01,
         2.83581608e-01, -2.30239961e-01, -9.79035590e-02,
        -2.05597435e-01, -1.86803983e-01, -1.98994418e-01,
        -2.92667291e-02, -6.35827176e-02,  1.15659751e-01,
        -1.70020803e-02,  6.18784338e-01,  1.50372991e+00],
       [-1.01640064e-01, -2.06518015e-01,  2.69695837e-01,
         1.73754566e-03, -4.94545312e-02, -6.58305552e-02,
         3.65184253e-03, -2.59751581e-01, -1.40703408e

array([-0.28474821, -0.31260391, -2.73439106])

In [254]:
test_output = pd.DataFrame(model.predict(X_test), index = X_test.index, columns = ['pred_Class'])
test_output.head()

pred_Class
42248         Eco
102458   Business
236           Eco
74922    Business
14632         Eco

In [255]:
test_output = test_output.merge(y_test, left_index = True, right_index = True)
test_output.head()
print('Percentage of correct predictions is ')
score_rounded = model.score(X_test, y_test)
print(score_rounded)

pred_Class     Class
42248         Eco  Business
102458   Business  Business
236           Eco       Eco
74922    Business  Business
14632         Eco       Eco

Percentage of correct predictions is 
0.7750019306510155


k-neighbors method

In [256]:
from sklearn.neighbors import KNeighborsClassifier

model = KNeighborsClassifier(n_neighbors=7)

# Fit the model on the training data
model.fit(X_train, y_train)

# Calculate the accuracy score on the training data
print(f"The model score with KNN")
model.score(X_train, y_train)

KNeighborsClassifier(n_neighbors=7)

The model score with KNN


0.8566063964320536

In [257]:
test_output = pd.DataFrame(model.predict(X_test), index = X_test.index, columns = ['pred_Class'])
test_output.head()

pred_Class
42248         Eco
102458   Business
236           Eco
74922    Business
14632         Eco

In [258]:
test_output = test_output.merge(y_test, left_index = True, right_index = True)
test_output.head()
print('Percentage of correct predictions is ')
score_rounded = model.score(X_test, y_test)
percentage = "{:.2f}".format(score_rounded * 100)
print(percentage)

pred_Class     Class
42248         Eco  Business
102458   Business  Business
236           Eco       Eco
74922    Business  Business
14632         Eco       Eco

Percentage of correct predictions is 
82.86


Multinomial Logistic Regression with Penalty

In [259]:
df= pd.read_csv('invistico_Airline.csv')
df.head()

satisfaction  Gender   Customer Type  Age   Type of Travel     Class  \
0    satisfied  Female  Loyal Customer   65  Personal Travel       Eco   
1    satisfied    Male  Loyal Customer   47  Personal Travel  Business   
2    satisfied  Female  Loyal Customer   15  Personal Travel       Eco   
3    satisfied  Female  Loyal Customer   60  Personal Travel       Eco   
4    satisfied  Female  Loyal Customer   70  Personal Travel       Eco   

   Flight Distance  Seat comfort  Departure/Arrival time convenient  \
0              265             0                                  0   
1             2464             0                                  0   
2             2138             0                                  0   
3              623             0                                  0   
4              354             0                                  0   

   Food and drink  ...  Online support  Ease of Online booking  \
0               0  ...               2                       3   
1               0  ...               2                       3   
2               0  ...               2                       2   
3               0  ...               3                       1   
4               0  ...               4                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 3                 0                 3                5   
1                 4                 4                 4                2   
2                 3                 3                 4                4   
3                 1                 0                 1                4   
4                 2                 0                 2                4   

   Cleanliness  Online boarding  Departure Delay in Minutes  \
0            3                2                           0   
1            3                2                         310   
2            4                2                           0   
3            1                3                           0   
4            2                5                           0   

   Arrival Delay in Minutes  
0                      0.00  
1                    305.00  
2                      0.00  
3                      0.00  
4                      0.00  

[5 rows x 23 columns]

In [260]:
df.dropna(inplace = True)

In [261]:
subset_data = df[df['Inflight entertainment'] == 0].drop(columns = ['satisfaction'])

subset_data.head()

len(subset_data)

Gender   Customer Type  Age   Type of Travel Class  Flight Distance  \
2   Female  Loyal Customer   15  Personal Travel   Eco             2138   
5     Male  Loyal Customer   30  Personal Travel   Eco             1894   
7     Male  Loyal Customer   10  Personal Travel   Eco             1812   
9     Male  Loyal Customer   22  Personal Travel   Eco             1556   
11  Female  Loyal Customer   34  Personal Travel   Eco             3633   

    Seat comfort  Departure/Arrival time convenient  Food and drink  \
2              0                                  0               0   
5              0                                  0               0   
7              0                                  0               0   
9              0                                  0               0   
11             0                                  0               0   

    Gate location  ...  Online support  Ease of Online booking  \
2               3  ...               2                       2   
5               3  ...               2                       2   
7               3  ...               2                       2   
9               3  ...               2                       2   
11              4  ...               2                       2   

    On-board service  Leg room service  Baggage handling  Checkin service  \
2                  3                 3                 4                4   
5                  5                 4                 5                5   
7                  3                 3                 4                5   
9                  2                 4                 5                3   
11                 3                 2                 5                2   

    Cleanliness  Online boarding  Departure Delay in Minutes  \
2             4                2                           0   
5             4                2                           0   
7             4                2                           0   
9             4                2                          30   
11            5                2                           0   

    Arrival Delay in Minutes  
2                       0.00  
5                       0.00  
7                       0.00  
9                      26.00  
11                      0.00  

[5 rows x 22 columns]

2968

In [262]:
from sklearn.preprocessing import OneHotEncoder

def get_ohe(df, col):
    ohe = OneHotEncoder(drop='first', handle_unknown='error', sparse_output=False, dtype='int')
    ohe.fit(df[[col]])
    temp_df = pd.DataFrame(data=ohe.transform(df[[col]]), columns=ohe.get_feature_names_out())
    df.drop(columns=[col], axis=1, inplace=True)
    df = pd.concat([df.reset_index(drop=True), temp_df], axis=1)
    return df

In [263]:
subset_data = get_ohe(subset_data, 'Gender')
subset_data = get_ohe(subset_data, 'Customer Type')
subset_data = get_ohe(subset_data, 'Type of Travel')
subset_data.head()

Age Class  Flight Distance  Seat comfort  \
0   15   Eco             2138             0   
1   30   Eco             1894             0   
2   10   Eco             1812             0   
3   22   Eco             1556             0   
4   34   Eco             3633             0   

   Departure/Arrival time convenient  Food and drink  Gate location  \
0                                  0               0              3   
1                                  0               0              3   
2                                  0               0              3   
3                                  0               0              3   
4                                  0               0              4   

   Inflight wifi service  Inflight entertainment  Online support  ...  \
0                      2                       0               2  ...   
1                      2                       0               2  ...   
2                      2                       0               2  ...   
3                      2                       0               2  ...   
4                      2                       0               2  ...   

   Leg room service  Baggage handling  Checkin service  Cleanliness  \
0                 3                 4                4            4   
1                 4                 5                5            4   
2                 3                 4                5            4   
3                 4                 5                3            4   
4                 2                 5                2            5   

   Online boarding  Departure Delay in Minutes  Arrival Delay in Minutes  \
0                2                           0                      0.00   
1                2                           0                      0.00   
2                2                           0                      0.00   
3                2                          30                     26.00   
4                2                           0                      0.00   

   Gender_Male  Customer Type_disloyal Customer  \
0            0                                0   
1            1                                0   
2            1                                0   
3            1                                0   
4            0                                0   

   Type of Travel_Personal Travel  
0                               1  
1                               1  
2                               1  
3                               1  
4                               1  

[5 rows x 22 columns]

In [264]:
X_train, X_test, y_train, y_test = train_test_split(subset_data.drop(columns = ['Class','Inflight entertainment']), subset_data['Class'], test_size=0.20, stratify=subset_data['Class'],random_state=50)

X_train
X_test
y_train
y_test
first_train_index = X_train.index[0]
print(f"The start index for train dataset is {first_train_index}")
first_test_index = X_test.index[0]
print(f"The start index for test dataset is {first_test_index}")

Age  Flight Distance  Seat comfort  Departure/Arrival time convenient  \
1227   22             1317             3                                  3   
1106    9             1625             2                                  4   
1541   28             1505             5                                  5   
1186   25             1322             3                                  0   
6      35             1766             0                                  1   
...   ...              ...           ...                                ...   
534    30             1582             1                                  0   
764     7             1865             1                                  4   
1304   45             1475             3                                  4   
2845   21             3857             0                                  5   
1973   23             2262             0                                  0   

      Food and drink  Gate location  Inflight wifi service  Online support  \
1227               0              4                      3               3   
1106               0              5                      2               4   
1541               0              3                      3               3   
1186               0              5                      2               2   
6                  0              1                      4               4   
...              ...            ...                    ...             ...   
534                0              3                      1               1   
764                0              1                      1               1   
1304               0              4                      4               4   
2845               0              2                      3               3   
1973               0              4                      5               5   

      Ease of Online booking  On-board service  Leg room service  \
1227                       3                 3                 1   
1106                       2                 4                 2   
1541                       3                 4                 4   
1186                       2                 3                 4   
6                          4                 3                 5   
...                      ...               ...               ...   
534                        1                 2                 4   
764                        1                 4                 4   
1304                       4                 5                 4   
2845                       3                 2                 3   
1973                       5                 4                 5   

      Baggage handling  Checkin service  Cleanliness  Online boarding  \
1227                 5                2            4                3   
1106                 4                4            5                2   
1541                 5                5            2                3   
1186                 4                5            4                2   
6                    2                3            2                4   
...                ...              ...          ...              ...   
534                  3                5            1                1   
764                  5                5            4                1   
1304                 1                2            2                4   
2845                 5                5            1                3   
1973                 5                4            5                5   

      Departure Delay in Minutes  Arrival Delay in Minutes  Gender_Male  \
1227                          18                     19.00            1   
1106                           0                      0.00            1   
1541                           0                      0.00            0   
1186                           0                      0.00            1   
6                              0                   

Age  Flight Distance  Seat comfort  Departure/Arrival time convenient  \
1868   22             1570             0                                  0   
1627   23             1329             0                                  0   
665    64             1428             0                                  5   
1597   23             2439             0                                  0   
847     7             2155             0                                  0   
...   ...              ...           ...                                ...   
529    64             1254             0                                  5   
2610   30             2186             0                                  0   
2271   25             2202             0                                  5   
149    17             1865             0                                  3   
944    33             3941             1                                  5   

      Food and drink  Gate location  Inflight wifi service  Online support  \
1868               0              3                      4               4   
1627               0              1                      5               2   
665                0              3                      1               2   
1597               0              1                      1               1   
847                0              2                      3               2   
...              ...            ...                    ...             ...   
529                0              5                      5               5   
2610               0              1                      4               4   
2271               0              3                      4               4   
149                0              3                      2               4   
944                0              3                      5               5   

      Ease of Online booking  On-board service  Leg room service  \
1868                       4                 3                 5   
1627                       5                 2                 4   
665                        1                 3                 5   
1597                       1                 4                 4   
847                        3                 4                 1   
...                      ...               ...               ...   
529                        5                 4                 5   
2610                       4                 4                 5   
2271                       4                 3                 4   
149                        2                 5                 3   
944                        5                 5                 5   

      Baggage handling  Checkin service  Cleanliness  Online boarding  \
1868                 4                5            4                4   
1627                 3                4            4                5   
665                  5                5            4                1   
1597                 4                5            5                1   
847                  1                5            4                3   
...                ...              ...          ...              ...   
529                  5                3            5                5   
2610                 3                1            4                4   
2271                 4                5            4                4   
149                  1                5            2                2   
944                  3                3            5                5   

      Departure Delay in Minutes  Arrival Delay in Minutes  Gender_Male  \
1868                           0                      0.00            1   
1627                           0                      0.00            0   
665                           48                     75.00            0   
1597                           0                      0.00            1   
847                           69                   

1227    Business
1106         Eco
1541         Eco
1186         Eco
6            Eco
          ...   
534          Eco
764          Eco
1304         Eco
2845    Eco Plus
1973    Business
Name: Class, Length: 2374, dtype: object

1868    Business
1627    Eco Plus
665          Eco
1597         Eco
847          Eco
          ...   
529          Eco
2610         Eco
2271    Business
149          Eco
944          Eco
Name: Class, Length: 594, dtype: object

The start index for train dataset is 1227
The start index for test dataset is 1868


In [265]:
if True:
    from sklearn.preprocessing import StandardScaler
    sc = StandardScaler()
    X_train = pd.DataFrame(sc.fit_transform(X_train), columns = X_train.columns, index = X_train.index)
    X_test = pd.DataFrame(sc.transform(X_test), columns = X_test.columns, index = X_test.index)
    X_train
    X_test
    y_train
    y_test

Age  Flight Distance  Seat comfort  Departure/Arrival time convenient  \
1227 -0.57            -0.89          1.49                               0.29   
1106 -1.45            -0.41          0.78                               0.78   
1541 -0.17            -0.60          2.91                               1.27   
1186 -0.37            -0.88          1.49                              -1.17   
6     0.30            -0.19         -0.64                              -0.68   
...    ...              ...           ...                                ...   
534  -0.03            -0.48          0.07                              -1.17   
764  -1.58            -0.04          0.07                               0.78   
1304  0.98            -0.64          1.49                               0.78   
2845 -0.64             3.06         -0.64                               1.27   
1973 -0.50             0.58         -0.64                              -1.17   

      Food and drink  Gate location  Inflight wifi service  Online support  \
1227           -0.14           0.95                  -0.03           -0.00   
1106           -0.14           1.79                  -0.77            0.73   
1541           -0.14           0.11                  -0.03           -0.00   
1186           -0.14           1.79                  -0.77           -0.73   
6              -0.14          -1.57                   0.71            0.73   
...              ...            ...                    ...             ...   
534            -0.14           0.11                  -1.50           -1.45   
764            -0.14          -1.57                  -1.50           -1.45   
1304           -0.14           0.95                   0.71            0.73   
2845           -0.14          -0.73                  -0.03           -0.00   
1973           -0.14           0.95                   1.44            1.45   

      Ease of Online booking  On-board service  Leg room service  \
1227                   -0.03             -0.26             -1.66   
1106                   -0.77              0.53             -0.91   
1541                   -0.03              0.53              0.60   
1186                   -0.77             -0.26              0.60   
6                       0.71             -0.26              1.36   
...                      ...               ...               ...   
534                    -1.50             -1.04              0.60   
764                    -1.50              0.53              0.60   
1304                    0.71              1.31              0.60   
2845                   -0.03             -1.04             -0.15   
1973                    1.44              0.53              1.36   

      Baggage handling  Checkin service  Cleanliness  Online boarding  \
1227              1.06            -1.05         0.24            -0.03   
1106              0.25             0.51         1.09            -0.77   
1541              1.06             1.28        -1.45            -0.03   
1186              0.25             1.28         0.24            -0.77   
6                -1.38            -0.27        -1.45             0.71   
...                ...              ...          ...              ...   
534              -0.56             1.28        -2.30            -1.50   
764               1.06             1.28         0.24            -1.50   
1304             -2.19            -1.05        -1.45             0.71   
2845              1.06             1.28        -2.30            -0.03   
1973              1.06             0.51         1.09             1.44   

      Departure Delay in Minutes  Arrival Delay in Minutes  Gender_Male  \
1227                        0.43                      0.47         0.85   
1106                       -0.46                     -0.46         0.85   
1541                       -0.46                     -0.46        -1.17   
1186                       -0.46                     -0.46         0.85   
6                          -0.46        

Age  Flight Distance  Seat comfort  Departure/Arrival time convenient  \
1868 -0.57            -0.50         -0.64                              -1.17   
1627 -0.50            -0.87         -0.64                              -1.17   
665   2.25            -0.72         -0.64                               1.27   
1597 -0.50             0.86         -0.64                              -1.17   
847  -1.58             0.41         -0.64                              -1.17   
...    ...              ...           ...                                ...   
529   2.25            -0.99         -0.64                               1.27   
2610 -0.03             0.46         -0.64                              -1.17   
2271 -0.37             0.49         -0.64                               1.27   
149  -0.91            -0.04         -0.64                               0.29   
944   0.17             3.19          0.07                               1.27   

      Food and drink  Gate location  Inflight wifi service  Online support  \
1868           -0.14           0.11                   0.71            0.73   
1627           -0.14          -1.57                   1.44           -0.73   
665            -0.14           0.11                  -1.50           -0.73   
1597           -0.14          -1.57                  -1.50           -1.45   
847            -0.14          -0.73                  -0.03           -0.73   
...              ...            ...                    ...             ...   
529            -0.14           1.79                   1.44            1.45   
2610           -0.14          -1.57                   0.71            0.73   
2271           -0.14           0.11                   0.71            0.73   
149            -0.14           0.11                  -0.77            0.73   
944            -0.14           0.11                   1.44            1.45   

      Ease of Online booking  On-board service  Leg room service  \
1868                    0.71             -0.26              1.36   
1627                    1.44             -1.04              0.60   
665                    -1.50             -0.26              1.36   
1597                   -1.50              0.53              0.60   
847                    -0.03              0.53             -1.66   
...                      ...               ...               ...   
529                     1.44              0.53              1.36   
2610                    0.71              0.53              1.36   
2271                    0.71             -0.26              0.60   
149                    -0.77              1.31             -0.15   
944                     1.44              1.31              1.36   

      Baggage handling  Checkin service  Cleanliness  Online boarding  \
1868              0.25             1.28         0.24             0.71   
1627             -0.56             0.51         0.24             1.44   
665               1.06             1.28         0.24            -1.50   
1597              0.25             1.28         1.09            -1.50   
847              -2.19             1.28         0.24            -0.03   
...                ...              ...          ...              ...   
529               1.06            -0.27         1.09             1.44   
2610             -0.56            -1.83         0.24             0.71   
2271              0.25             1.28         0.24             0.71   
149              -2.19             1.28        -1.45            -0.77   
944              -0.56            -0.27         1.09             1.44   

      Departure Delay in Minutes  Arrival Delay in Minutes  Gender_Male  \
1868                       -0.46                     -0.46         0.85   
1627                       -0.46                     -0.46        -1.17   
665                         1.91                      3.21        -1.17   
1597                       -0.46                     -0.46         0.85   
847                         2.94        

1227    Business
1106         Eco
1541         Eco
1186         Eco
6            Eco
          ...   
534          Eco
764          Eco
1304         Eco
2845    Eco Plus
1973    Business
Name: Class, Length: 2374, dtype: object

1868    Business
1627    Eco Plus
665          Eco
1597         Eco
847          Eco
          ...   
529          Eco
2610         Eco
2271    Business
149          Eco
944          Eco
Name: Class, Length: 594, dtype: object

In [266]:
model = LogisticRegression(fit_intercept = True, solver='lbfgs', multi_class = 'ovr', penalty = None)
# model = LogisticRegression(fit_intercept = True, solver='lbfgs', multi_class = 'ovr', penalty = 'none', max_iter = 1000)
# model = LogisticRegression(fit_intercept = True, solver='lbfgs', multi_class = 'multinomial', penalty = 'none', max_iter = 1000)
# model = LogisticRegression(fit_intercept = True, solver='newton-cg', multi_class = 'multinomial', penalty = 'none', max_iter = 10000)

# While using multiclass case do multi_class = 'ovr' or 'auto'; can also try other solvers
# While doing regularization, use penalty = 'l2' and also C = 10.0 (need to try other values too)

model.fit(X_train, y_train) 

# The following gives the mean accuracy on the given data and labels
train_score = model.score(X_train, y_train) 
test_score = model.score(X_test, y_test)
train_score_percentage = "{:.2f}".format(train_score * 100)
test_score_percentage = "{:.2f}".format(test_score * 100)
print(f"Accuracy for train dataset is: {train_score_percentage} %")
print(f"Accuracy value for test dataset is: {test_score_percentage} %")

# This is the coefficient Beta_1, ..., Beta_7
model.coef_

# This is the coefficient Beta_0
model.intercept_

LogisticRegression(multi_class='ovr', penalty=None)

Accuracy for train dataset is: 78.10 %
Accuracy value for test dataset is: 75.76 %


array([[-0.47621213,  0.50458457, -0.06448408, -0.33471057,  0.0214228 ,
         0.136524  , -0.02677323,  0.08269893, -0.02677323,  0.19339392,
         0.18297614,  0.44785498,  0.17571178,  0.3999089 , -0.02677323,
        -0.17356831,  0.24310981,  0.05412453,  0.83162775, -0.61930582],
       [ 0.11940852, -0.26717072,  0.09330552,  0.24956706, -0.03542464,
         0.0216197 ,  0.00890864,  0.0291108 ,  0.00890864, -0.17001247,
        -0.09845438, -0.24200546, -0.14589332, -0.23354806,  0.00890864,
         0.08068188, -0.06336826, -0.02415577,  0.07606072,  0.92910376],
       [ 0.10242211, -0.039159  , -0.06795014,  0.05612588, -0.75621926,
        -0.22784554,  0.02064203, -0.15553278,  0.02064203,  0.00927705,
        -0.00599129, -0.05912747,  0.0382984 , -0.02452526,  0.02064203,
        -0.00153895, -0.12096803, -0.04915172, -1.36270782, -0.89377005]])

array([-2.23536051,  0.9030872 , -2.52839323])

In [267]:
model = LogisticRegression(penalty='l1', solver='liblinear', multi_class='ovr', C=0.1)
model.fit(X_train, y_train)

# Prediction accuracy on the train set
train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)
train_score_percentage = "{:.2f}".format(train_score * 100)
test_score_percentage = "{:.2f}".format(test_score * 100)
print(f"Accuracy for train dataset is: {train_score_percentage} %")
print(f"Accuracy value for test dataset is: {test_score_percentage} %")

X=subset_data.drop(columns = ['Class','Inflight entertainment'])
zero_coef_features = X.columns[model.coef_[0] == 0]
print("Features with nearly zero coefficients:", zero_coef_features)

LogisticRegression(C=0.1, multi_class='ovr', penalty='l1', solver='liblinear')

Accuracy for train dataset is: 77.72 %
Accuracy value for test dataset is: 76.43 %
Features with nearly zero coefficients: Index(['Seat comfort', 'Food and drink', 'Inflight wifi service',
       'Online support', 'Ease of Online booking', 'Online boarding',
       'Departure Delay in Minutes'],
      dtype='object')


The test and train accuracy are almost same as the previous model.It hasn't improved much.

In [268]:
model = LogisticRegression(penalty='l2', solver='liblinear', multi_class='ovr', C=0.1)
model.fit(X_train, y_train)

# Prediction accuracy on the train set
train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)
train_score_percentage = "{:.2f}".format(train_score * 100)
test_score_percentage = "{:.2f}".format(test_score * 100)
print(f"Accuracy for train dataset is: {train_score_percentage} %")
print(f"Accuracy value for test dataset is: {test_score_percentage} %")

X=subset_data.drop(columns = ['Class','Inflight entertainment'])
zero_coef_features = X.columns[model.coef_[0] == 0]
print("Features with nearly zero coefficients:", zero_coef_features)

LogisticRegression(C=0.1, multi_class='ovr', solver='liblinear')

Accuracy for train dataset is: 78.14 %
Accuracy value for test dataset is: 76.26 %
Features with nearly zero coefficients: Index([], dtype='object')
